src/pull.rs

```diff
@@ -203,6 +204,152 @@ impl<'a> PullClient<'a> {
+                    .map_err(|e| anyhow!("failed to async pull blob {}", e.to_string()))?;
+
+                self.async_handle_layer(
+                    layer,
+                    diff_ids[i].clone(),
+                    decrypt_config,
+                    layer_reader,
+                    ms,
+                )
+                .await
+                .map_err(|e| anyhow!("failed to async handle layer: {:?}", e))
+            }
+        });
+
+        let layer_metas = future::try_join_all(layer_metas).await?;
+
+        Ok(layer_metas)
+    }
+
+    async fn async_handle_layer(
+        &self,
+        layer: OciDescriptor,
+        diff_id: String,
+        decrypt_config: &Option<&str>,
+        layer_reader: (impl tokio::io::AsyncRead + Unpin),
+        ms: Arc<Mutex<MetaStore>>,
+    ) -> Result<LayerMeta> {
+        let layer_db = &ms.lock().await.layer_db;
+
+        if let Some(layer_meta) = layer_db.get(&layer.digest) {
+            return Ok(layer_meta.clone());
+        }
+
+        let mut layer_meta = LayerMeta::default();
+        let mut media_type_str: &str = layer.media_type.as_str();
+
+        let store_path = format!(
+            "{}/{}",
+            self.data_dir.display(),
+            &layer.digest.to_string().replace(':', "_")
+        );
+
+        let destination = Path::new(&store_path);
+
+        let decryptor = Decryptor::from_media_type(&layer.media_type);
+
+        let uncompressed_digest;
+
+        if decryptor.is_encrypted() {
+            if let Some(dc) = decrypt_config {
+                let decrypt_key = decryptor
+                    .get_decrypt_key(&layer, dc)
+                    .await
+                    .map_err(|e| anyhow!("failed to get decrypt key {}", e.to_string()))?;
+
+                let plaintext_layer = decryptor
+                    .async_get_plaintext_layer(layer_reader, &layer, &decrypt_key)
+                    .map_err(|e| anyhow!("failed to async_get_plaintext_layer: {:?}", e))?;
+
+                layer_meta.encrypted = true;
+                media_type_str = decryptor.media_type.as_str();
+                uncompressed_digest = self
+                    .async_decompress_unpack_layer(
+                        plaintext_layer,
+                        &diff_id,
+                        media_type_str,
+                        destination,
+                    )
+                    .await?;
+            } else {
+                return Err(anyhow!(ERR_NO_DECRYPT_CFG));
+            }
+        } else {
+            uncompressed_digest = self
+                .async_decompress_unpack_layer(layer_reader, &diff_id, media_type_str, destination)
+                .await?;
+        }
+
+        layer_meta.compressed_digest = layer.digest.clone();
+        layer_meta.uncompressed_digest = uncompressed_digest;
+
+        // uncompressed digest should equal to the diff_ids in image_config.
+        if layer_meta.uncompressed_digest != diff_id {
+            println!(
```

Comment:
Not related to this PR, too. I found the codebase also includes another `println!`. Maybe we can use `log::info` instead of `println` macro, for `log` might have better extensibility?